File: src/core/baidu.ts
import fetch from 'node-fetch';
import { Translator } from './translator';
import crypto from 'crypto';

export class BaiduFanyi implements Translator {
  name!:string;

  targetLang: string;
  appid: string;
  secret: string;

  constructor(targetLang: string, appid: string, secret: string) {
    this.targetLang = targetLang;
    this.appid = appid;
    this.secret = secret;
  }

  async translate(text: string): Promise<string> {
    let url = new URL('https://fanyi-api.baidu.com/api/trans/vip/translate');

    let salt = String(Math.random());
    let sign = crypto.createHash('md5').update(this.appid + text + salt + this.secret).digest("hex");

    const params = new URLSearchParams();
    params.append('q', text);
    params.append('from', 'ja');
    params.append('to', this.targetLang.split('-')[0]);
    params.append('appid', this.appid);
    params.append('salt', salt);
    params.append('sign', sign);

    console.log(text);

    let response = await fetch(url, {
      method: 'POST',
      body: params,
      headers: {
      }
    });
    let body = await response.json();
    if (!response.ok) {
      throw new Error(body.error.message);
    }
    return body.trans_result[0].dst;
  }
}
BaiduFanyi.prototype.name = 'Baidu';

File: server/src/translator/baidu.rs
use super::Translator;
use async_trait::async_trait;

pub struct BaiduTranslator {
    appid: String,
    secret: String,
    target_lang: String,
}

impl BaiduTranslator {
    pub fn new(appid: String, secret: String, target_lang: String) -> Self {
        Self {
            appid,
            secret,
            target_lang,
        }
    }
}

#[async_trait]
impl Translator for BaiduTranslator {
    fn name(&self) -> &'static str {
        "Baidu"
    }

    async fn translate(&self, text: &str) -> anyhow::Result<String> {
        const API_URL: &str = "https://fanyi-api.baidu.com/api/trans/vip/translate";

        let salt = format!("{:x}", rand::random::<u32>());
        let sign = format!(
            "{:x}",
            md5::compute(&format!("{}{}{}{}", self.appid, text, salt, self.secret))
        );

        #[derive(serde::Deserialize)]
        struct TransResult {
            dst: String,
        }

        #[derive(serde::Deserialize)]
        #[allow(dead_code)]
        struct Response {
            trans_result: Vec<TransResult>,
        }

        let client = reqwest::Client::new();
        let body: Response = client
            .post(API_URL)
            .form(&[
                ("q", text),
                ("from", "jp"),
                ("to", &*self.target_lang),
                ("appid", &*self.appid),
                ("salt", &*salt),
                ("sign", &*sign),
            ])
            .send()
            .await?
            .json()
            .await?;

        Ok(body
            .trans_result
            .into_iter()
            .map(|v| v.dst)
            .collect::<Vec<_>>()
            .join(""))
    }
}
